File: apps/mobile/modules/vexl-local-transport/android/src/main/java/expo/modules/vexllocaltransport/VexlLocalTransportModule.kt
package expo.modules.vexllocaltransport

import android.util.Base64
import androidx.core.os.bundleOf
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.IOException
import java.net.Inet4Address
import java.net.InetSocketAddress
import java.net.NetworkInterface
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketException
import java.net.SocketTimeoutException
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.atomic.AtomicBoolean

// Read loop chunk size. Must stay in sync with the data chunk plaintext limit
// declared in packages/domain/src/general/deviceMigration/limits.ts.
private const val MAX_RECEIVE_CHUNK_BYTES = 64 * 1024

private class ListenerAlreadyActiveException :
  CodedException("A local transport listener is already active")

private class ListenerStartFailedException :
  CodedException("The local transport listener could not be started")

private class NoActiveListenerException :
  CodedException("No local transport listener is active")

private class InvalidPortException :
  CodedException("Port must be an integer between 1 and 65535")

private class ConnectFailedException :
  CodedException("The local transport connection could not be established")

private class ConnectTimedOutException :
  CodedException("The local transport connection attempt timed out or was cancelled")

private class ConnectionNotFoundException :
  CodedException("No open local transport connection with the given id exists")

private class InvalidBase64PayloadException :
  CodedException("The payload is not valid base64")

private class SendFailedException :
  CodedException("Sending on the local transport connection failed")

/**
 * Minimal local-network TCP transport used exclusively by device migration.
 *
 * Deliberately thin: framing, limits, timeouts and the protocol state machine
 * live in the javascript protocol layer. There is no TLS at this layer — the
 * application-layer secretstream encrypts every byte. No mDNS/NSD is used;
 * the pairing QR carries the reachable endpoints.
 *
 * Privacy: nothing handled here (payloads, hosts, ports) may ever be logged
 * or reported. Errors crossing the bridge carry only the static messages of
 * the typed exceptions above.
 */
class VexlLocalTransportModule : Module() {
  private val stateLock = Any()
  private var serverSocket: ServerSocket? = null
  private var hasAcceptedInboundConnection = false
  private var destroyed = false
  private val connections = ConcurrentHashMap<String, ConnectionHandle>()

  override fun definition() = ModuleDefinition {
    Name("VexlLocalTransport")

    Events("onConnectionAccepted", "onData", "onConnectionClosed", "onListenerFailed")

    AsyncFunction("startListener") { promise: Promise ->
      synchronized(stateLock) {
        if (serverSocket != null) {
          promise.reject(ListenerAlreadyActiveException())
          return@AsyncFunction
        }
        val socket = try {
          // Port 0 — the system assigns an ephemeral port bound on all
          // interfaces (wildcard address).
          ServerSocket(0)
        } catch (exception: IOException) {
          promise.reject(ListenerStartFailedException())
          return@AsyncFunction
        }
        serverSocket = socket
        hasAcceptedInboundConnection = false
        Thread { acceptLoop(socket) }.apply {
          isDaemon = true
          name = "vexl-local-transport-accept"
          start()
        }
        promise.resolve(bundleOf("port" to socket.localPort))
      }
    }

    AsyncFunction("getLocalEndpointCandidates") { promise: Promise ->
      val port = synchronized(stateLock) { serverSocket?.localPort }
      if (port == null) {
        promise.reject(NoActiveListenerException())
        return@AsyncFunction
      }
      promise.resolve(
        nonLoopbackIpv4Addresses().map { bundleOf("host" to it, "port" to port) }
      )
    }

    AsyncFunction("stopListener") { promise: Promise ->
      stopListener()
      promise.resolve(null)
    }

    AsyncFunction("connect") { host: String, port: Int, timeoutMs: Int, promise: Promise ->
      if (port < 1 || port > 65535) {
        promise.reject(InvalidPortException())
        return@AsyncFunction
      }
      // A dedicated thread so a slow connect attempt cannot block other
      // module calls. The blocking connect itself enforces the timeout.
      Thread {
        val socket = Socket()
        try {
          socket.tcpNoDelay = true
          socket.connect(InetSocketAddress(host, port), timeoutMs.coerceAtLeast(0))
          val handle = registerConnection(socket)
          promise.resolve(bundleOf("connectionId" to handle.id))
        } catch (exception: SocketTimeoutException) {
          closeQuietly(socket)
          promise.reject(ConnectTimedOutException())
        } catch (exception: Exception) {
          closeQuietly(socket)
          promise.reject(ConnectFailedException())
        }
      }.apply {
        isDaemon = true
        name = "vexl-local-transport-connect"
        start()
      }
    }

    AsyncFunction("send") { connectionId: String, dataBase64: String, promise: Promise ->
      val handle = connections[connectionId]
      if (handle == null) {
        promise.reject(ConnectionNotFoundException())
        return@AsyncFunction
      }
      val data = try {
        Base64.decode(dataBase64, Base64.DEFAULT)
      } catch (exception: IllegalArgumentException) {
        promise.reject(InvalidBase64PayloadException())
        return@AsyncFunction
      }
      handle.send(data, promise)
    }

    AsyncFunction("closeConnection") { connectionId: String, promise: Promise ->
      // Idempotent — closing an unknown or already closed connection is a
      // no-op success.
      connections[connectionId]?.close(locally = true)
      promise.resolve(null)
    }

    OnDestroy {
      synchronized(stateLock) {
        destroyed = true
      }
      stopListener()
      val openConnections = connections.values.toList()
      connections.clear()
      for (handle in openConnections) {
        handle.shutdownSilently()
      }
    }
  }

  private fun stopListener() {
    val socket = synchronized(stateLock) {
      val current = serverSocket
      serverSocket = null
      current
    }
    if (socket != null) {
      try {
        socket.close()
      } catch (exception: IOException) {
        // Already closed — nothing else to release.
      }
    }
  }

  private fun acceptLoop(listeningSocket: ServerSocket) {
    while (true) {
      val socket = try {
        listeningSocket.accept()
      } catch (exception: IOException) {
        val wasStopped = synchronized(stateLock) {
          destroyed || serverSocket !== listeningSocket
        }
        if (!wasStopped) {
          synchronized(stateLock) {
            if (serverSocket === listeningSocket) serverSocket = null
          }
          closeQuietly(listeningSocket)
          sendEvent("onListenerFailed", bundleOf("reason" to "error"))
        }
        return
      }

      val acceptThisSocket = synchronized(stateLock) {
        if (hasAcceptedInboundConnection) {
          false
        } else {
          hasAcceptedInboundConnection = true
          true
        }
      }
      if (!acceptThisSocket) {
        // At most one inbound peer is ever accepted per listener session.
        // Later connection attempts are refused outright and never surfaced
        // to javascript. The typescript wrapper enforces this too.
        closeQuietly(socket)
        continue
      }

      try {
        socket.tcpNoDelay = true
      } catch (exception: SocketException) {
        // Best effort only.
      }
      val handle = registerConnection(socket)
      sendEvent(
        "onConnectionAccepted",
        bundleOf(
          "connectionId" to handle.id,
          "remoteHost" to (socket.inetAddress?.hostAddress ?: "")
        )
      )
    }
  }

  private fun registerConnection(socket: Socket): ConnectionHandle {
    val handle = ConnectionHandle(UUID.randomUUID().toString(), socket)
    connections[handle.id] = handle
    handle.startReadLoop()
    return handle
  }

  private fun nonLoopbackIpv4Addresses(): List<String> {
    val addresses = mutableListOf<String>()
    try {
      for (networkInterface in NetworkInterface.getNetworkInterfaces()) {
        if (!networkInterface.isUp || networkInterface.isLoopback) continue
        for (address in networkInterface.inetAddresses) {
          if (address is Inet4Address && !address.isLoopbackAddress) {
            val host = address.hostAddress
            if (host != null && !addresses.contains(host)) {
              addresses.add(host)
            }
          }
        }
      }
    } catch (exception: SocketException) {
      // No interfaces readable — return what was collected so far.
    }
    return addresses
  }

  private fun closeQuietly(socket: Socket) {
    try {
      socket.close()
    } catch (exception: IOException) {
      // Already closed.
    }
  }

  private fun closeQuietly(socket: ServerSocket) {
    try {
      socket.close()
    } catch (exception: IOException) {
      // Already closed.
    }
  }

  private inner class ConnectionHandle(val id: String, val socket: Socket) {
    private val closed = AtomicBoolean(false)
    private val closedLocally = AtomicBoolean(false)
    private val silent = AtomicBoolean(false)
    // Serializes writes and provides backpressure — a send promise resolves
    // only after the blocking write handed the bytes to the OS buffer.
    private val writeExecutor: ExecutorService =
      Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable).apply {
          isDaemon = true
          name = "vexl-local-transport-write"
        }
      }

    fun startReadLoop() {
      Thread {
        val buffer = ByteArray(MAX_RECEIVE_CHUNK_BYTES)
        try {
          val input = socket.getInputStream()
          while (true) {
            val readCount = input.read(buffer)
            if (readCount == -1) {
              close(locally = false, reason = "closedByPeer")
              return@Thread
            }
            if (readCount > 0 && !closed.get()) {
              sendEvent(
                "onData",
                bundleOf(
                  "connectionId" to id,
                  "dataBase64" to Base64.encodeToString(buffer, 0, readCount, Base64.NO_WRAP)
                )
              )
            }
          }
        } catch (exception: IOException) {
          close(
            locally = false,
            reason = if (closedLocally.get()) "cancelled" else "error"
          )
        }
      }.apply {
        isDaemon = true
        name = "vexl-local-transport-read"
        start()
      }
    }

    fun send(data: ByteArray, promise: Promise) {
      try {
        writeExecutor.execute {
          try {
            val output = socket.getOutputStream()
            output.write(data)
            output.flush()
            promise.resolve(null)
          } catch (exception: IOException) {
            promise.reject(SendFailedException())
          }
        }
      } catch (exception: RejectedExecutionException) {
        // The connection was closed and its write executor shut down.
        promise.reject(ConnectionNotFoundException())
      }
    }

    fun close(locally: Boolean, reason: String = "cancelled") {
      if (locally) {
        closedLocally.set(true)
      }
      if (!closed.compareAndSet(false, true)) return
      connections.remove(id)
      writeExecutor.shutdown()
      closeQuietly(socket)
      if (!silent.get()) {
        sendEvent(
          "onConnectionClosed",
          bundleOf("connectionId" to id, "reason" to reason)
        )
      }
    }

    /** Close during module destruction — javascript is gone, emit nothing. */
    fun shutdownSilently() {
      silent.set(true)
      closedLocally.set(true)
      close(locally = true)
    }
  }
}

File: apps/mobile/modules/vexl-local-transport/ios/VexlLocalTransportModule.swift
import ExpoModulesCore
import Foundation
import Network

// Read loop chunk size. Must stay in sync with the data chunk plaintext limit
// declared in packages/domain/src/general/deviceMigration/limits.ts.
private let maxReceiveChunkBytes = 64 * 1024

internal final class ListenerAlreadyActiveException: Exception {
  override var reason: String {
    "A local transport listener is already active"
  }
}

internal final class ListenerStartFailedException: Exception {
  override var reason: String {
    "The local transport listener could not be started"
  }
}

internal final class NoActiveListenerException: Exception {
  override var reason: String {
    "No local transport listener is active"
  }
}

internal final class InvalidPortException: Exception {
  override var reason: String {
    "Port must be an integer between 1 and 65535"
  }
}

internal final class ConnectFailedException: Exception {
  override var reason: String {
    "The local transport connection could not be established"
  }
}

internal final class ConnectTimedOutException: Exception {
  override var reason: String {
    "The local transport connection attempt timed out or was cancelled"
  }
}

internal final class ConnectionNotFoundException: Exception {
  override var reason: String {
    "No open local transport connection with the given id exists"
  }
}

internal final class InvalidBase64PayloadException: Exception {
  override var reason: String {
    "The payload is not valid base64"
  }
}

internal final class SendFailedException: Exception {
  override var reason: String {
    "Sending on the local transport connection failed"
  }
}

/**
 * Minimal local-network TCP transport used exclusively by device migration.
 *
 * Deliberately thin: framing, limits, timeouts and the protocol state machine
 * live in the javascript protocol layer. There is no TLS at this layer — the
 * application-layer secretstream encrypts every byte. No mDNS/Bonjour is used;
 * the pairing QR carries the reachable endpoints.
 *
 * Privacy: nothing handled here (payloads, hosts, ports) may ever be logged
 * or reported. Errors crossing the bridge carry only the static messages of
 * the typed exceptions above.
 */
public class VexlLocalTransportModule: Module {
  // All mutable state is confined to this serial queue. The listener and all
  // connections are started on it, so their handlers also run on it.
  private let transportQueue = DispatchQueue(label: "it.vexl.local-transport")

  private var listener: NWListener?
  private var hasAcceptedInboundConnection = false
  private var connections: [String: NWConnection] = [:]
  // Connections javascript knows about (accepted event sent or connect
  // resolved). Only these produce onConnectionClosed events.
  private var announcedConnectionIds: Set<String> = []

  public func definition() -> ModuleDefinition {
    Name("VexlLocalTransport")

    Events("onConnectionAccepted", "onData", "onConnectionClosed", "onListenerFailed")

    AsyncFunction("startListener") { (promise: Promise) in
      self.transportQueue.async {
        self.startListenerOnQueue(promise: promise)
      }
    }

    AsyncFunction("getLocalEndpointCandidates") { (promise: Promise) in
      self.transportQueue.async {
        self.getLocalEndpointCandidatesOnQueue(promise: promise)
      }
    }

    AsyncFunction("stopListener") { (promise: Promise) in
      self.transportQueue.async {
        self.stopListenerOnQueue()
        promise.resolve(nil)
      }
    }

    AsyncFunction("connect") { (host: String, port: Int, timeoutMs: Int, promise: Promise) in
      self.transportQueue.async {
        self.connectOnQueue(host: host, port: port, timeoutMs: timeoutMs, promise: promise)
      }
    }

    AsyncFunction("send") { (connectionId: String, dataBase64: String, promise: Promise) in
      self.transportQueue.async {
        self.sendOnQueue(connectionId: connectionId, dataBase64: dataBase64, promise: promise)
      }
    }

    AsyncFunction("closeConnection") { (connectionId: String, promise: Promise) in
      self.transportQueue.async {
        self.notifyClosedOnQueue(connectionId: connectionId, reason: "cancelled")
        promise.resolve(nil)
      }
    }

    OnDestroy {
      self.transportQueue.async {
        self.teardownOnQueue()
      }
    }
  }

  // MARK: - Listener

  private func startListenerOnQueue(promise: Promise) {
    guard listener == nil else {
      promise.reject(ListenerAlreadyActiveException())
      return
    }

    let newListener: NWListener
    do {
      // No port requested — the system assigns an ephemeral port bound on all
      // interfaces.
      newListener = try NWListener(using: makeTcpParameters())
    } catch {
      promise.reject(ListenerStartFailedException())
      return
    }

    listener = newListener
    hasAcceptedInboundConnection = false
    var promiseSettled = false

    newListener.stateUpdateHandler = { [weak self] state in
      guard let self else {
        return
      }
      switch state {
      case .ready:
        if !promiseSettled {
          promiseSettled = true
          if let port = newListener.port?.rawValue {
            promise.resolve(["port": Int(port)])
          } else {
            self.cleanupListenerOnQueue(newListener)
            promise.reject(ListenerStartFailedException())
          }
        }
      case .failed:
        self.cleanupListenerOnQueue(newListener)
        if !promiseSettled {
          promiseSettled = true
          promise.reject(ListenerStartFailedException())
        } else {
          self.sendEvent("onListenerFailed", ["reason": "error"])
        }
      case .cancelled:
        if !promiseSettled {
          promiseSettled = true
          promise.reject(ListenerStartFailedException())
        }
      default:
        break
      }
    }

    newListener.newConnectionHandler = { [weak self] connection in
      self?.handleInboundConnectionOnQueue(connection)
    }

    newListener.start(queue: transportQueue)
  }

  private func cleanupListenerOnQueue(_ staleListener: NWListener) {
    staleListener.cancel()
    if listener === staleListener {
      listener = nil
    }
  }

  private func stopListenerOnQueue(clearHandlers: Bool = false) {
    if let activeListener = listener {
      listener = nil
      if clearHandlers {
        // Only during module destruction — javascript is gone. Otherwise the
        // state handler must stay attached so a startListener promise that is
        // still pending settles when the cancellation lands.
        activeListener.stateUpdateHandler = nil
      }
      activeListener.newConnectionHandler = nil
      activeListener.cancel()
    }
  }

  private func handleInboundConnectionOnQueue(_ connection: NWConnection) {
    if hasAcceptedInboundConnection {
      // At most one inbound peer is ever accepted per listener session.
      // Later connection attempts are refused outright and never surfaced
      // to javascript. The typescript wrapper enforces this too.
      connection.cancel()
      return
    }
    hasAcceptedInboundConnection = true

    let connectionId = UUID().uuidString
    connections[connectionId] = connection
    var becameReady = false

    connection.stateUpdateHandler = { [weak self] state in
      guard let self else {
        return
      }
      switch state {
      case .ready:
        becameReady = true
        self.announcedConnectionIds.insert(connectionId)
        self.sendEvent("onConnectionAccepted", [
          "connectionId": connectionId,
          "remoteHost": Self.remoteHostDescription(of: connection),
        ])
        self.startReceiveLoopOnQueue(connectionId: connectionId, connection: connection)
      case .failed:
        if !becameReady {
          // The only inbound slot was never actually established — allow a
          // later attempt from the peer instead of bricking the pairing.
          self.hasAcceptedInboundConnection = false
        }
        self.notifyClosedOnQueue(connectionId: connectionId, reason: "error")
      case .cancelled:
        self.notifyClosedOnQueue(connectionId: connectionId, reason: "cancelled")
      default:
        break
      }
    }

    connection.start(queue: transportQueue)
  }

  private func getLocalEndpointCandidatesOnQueue(promise: Promise) {
    guard let activeListener = listener, let port = activeListener.port?.rawValue else {
      promise.reject(NoActiveListenerException())
      return
    }
    let candidates: [[String: Any]] = Self.nonLoopbackIpv4Addresses().map { address in
      ["host": address, "port": Int(port)]
    }
    promise.resolve(candidates)
  }

  // MARK: - Connections

  private func connectOnQueue(host: String, port: Int, timeoutMs: Int, promise: Promise) {
    guard port >= 1, port <= 65535 else {
      promise.reject(InvalidPortException())
      return
    }
    guard !host.isEmpty, let nwPort = NWEndpoint.Port(rawValue: UInt16(port)) else {
      promise.reject(ConnectFailedException())
      return
    }

    let connection = NWConnection(
      host: NWEndpoint.Host(host),
      port: nwPort,
      using: makeTcpParameters()
    )
    let connectionId = UUID().uuidString
    var promiseSettled = false

    connection.stateUpdateHandler = { [weak self] state in
      guard let self else {
        return
      }
      switch state {
      case .ready:
        if !promiseSettled {
          promiseSettled = true
          self.connections[connectionId] = connection
          self.announcedConnectionIds.insert(connectionId)
          self.startReceiveLoopOnQueue(connectionId: connectionId, connection: connection)
          promise.resolve(["connectionId": connectionId])
        }
      case .failed:
        if !promiseSettled {
          promiseSettled = true
          connection.cancel()
          promise.reject(ConnectFailedException())
        } else {
          self.notifyClosedOnQueue(connectionId: connectionId, reason: "error")
        }
      case .cancelled:
        if !promiseSettled {
          promiseSettled = true
          promise.reject(ConnectTimedOutException())
        } else {
          self.notifyClosedOnQueue(connectionId: connectionId, reason: "cancelled")
        }
      default:
        break
      }
    }

    transportQueue.asyncAfter(deadline: .now() + .milliseconds(max(timeoutMs, 0))) {
      if !promiseSettled {
        // Cancelling moves the connection to .cancelled, which rejects the
        // promise with a timeout above.
        connection.cancel()
      }
    }

    connection.start(queue: transportQueue)
  }

  private func sendOnQueue(connectionId: String, dataBase64: String, promise: Promise) {
    guard let connection = connections[connectionId] else {
      promise.reject(ConnectionNotFoundException())
      return
    }
    guard let data = Data(base64Encoded: dataBase64) else {
      promise.reject(InvalidBase64PayloadException())
      return
    }
    // .contentProcessed provides backpressure — the promise resolves only
    // after the stack has taken ownership of the bytes.
    connection.send(content: data, completion: .contentProcessed { error in
      if error != nil {
        promise.reject(SendFailedException())
      } else {
        promise.resolve(nil)
      }
    })
  }

  private func startReceiveLoopOnQueue(connectionId: String, connection: NWConnection) {
    connection.receive(
      minimumIncompleteLength: 1,
      maximumLength: maxReceiveChunkBytes
    ) { [weak self] data, _, isComplete, error in
      guard let self else {
        return
      }
      // Runs on transportQueue because the connection was started on it.
      guard self.connections[connectionId] === connection else {
        // The connection was closed locally while a receive was in flight.
        return
      }
      if let data, !data.isEmpty {
        self.sendEvent("onData", [
          "connectionId": connectionId,
          "dataBase64": data.base64EncodedString(),
        ])
      }
      if error != nil {
        self.notifyClosedOnQueue(connectionId: connectionId, reason: "error")
        return
      }
      if isComplete {
        self.notifyClosedOnQueue(connectionId: connectionId, reason: "closedByPeer")
        return
      }
      self.startReceiveLoopOnQueue(connectionId: connectionId, connection: connection)
    }
  }

  private func notifyClosedOnQueue(connectionId: String, reason: String) {
    guard let connection = connections.removeValue(forKey: connectionId) else {
      return
    }
    connection.cancel()
    if announcedConnectionIds.remove(connectionId) != nil {
      sendEvent("onConnectionClosed", [
        "connectionId": connectionId,
        "reason": reason,
      ])
    }
  }

  private func teardownOnQueue() {
    stopListenerOnQueue(clearHandlers: true)
    let openConnections = connections
    connections.removeAll()
    announcedConnectionIds.removeAll()
    for (_, connection) in openConnections {
      connection.stateUpdateHandler = nil
      connection.cancel()
    }
  }

  // MARK: - Helpers

  private func makeTcpParameters() -> NWParameters {
    let tcpOptions = NWProtocolTCP.Options()
    tcpOptions.noDelay = true
    // No TLS at this layer — the application-layer secretstream encrypts.
    return NWParameters(tls: nil, tcp: tcpOptions)
  }

  private static func remoteHostDescription(of connection: NWConnection) -> String {
    guard case .hostPort(let host, _) = connection.endpoint else {
      return ""
    }
    switch host {
    case .ipv4(let address):
      return "\(address)"
    case .ipv6(let address):
      return "\(address)"
    case .name(let name, _):
      return name
    @unknown default:
      return ""
    }
  }

  private static func nonLoopbackIpv4Addresses() -> [String] {
    var addresses: [String] = []
    var interfaceList: UnsafeMutablePointer<ifaddrs>?
    guard getifaddrs(&interfaceList) == 0, let firstInterface = interfaceList else {
      return []
    }
    defer {
      freeifaddrs(interfaceList)
    }

    var pointer: UnsafeMutablePointer<ifaddrs>? = firstInterface
    while let current = pointer {
      let interface = current.pointee
      pointer = interface.ifa_next

      guard let interfaceAddress = interface.ifa_addr else {
        continue
      }
      let flags = Int32(bitPattern: interface.ifa_flags)
      guard (flags & IFF_UP) != 0, (flags & IFF_LOOPBACK) == 0 else {
        continue
      }
      guard interfaceAddress.pointee.sa_family == UInt8(AF_INET) else {
        continue
      }

      var hostBuffer = [CChar](repeating: 0, count: Int(NI_MAXHOST))
      let result = getnameinfo(
        interfaceAddress,
        socklen_t(interfaceAddress.pointee.sa_len),
        &hostBuffer,
        socklen_t(hostBuffer.count),
        nil,
        0,
        NI_NUMERICHOST
      )
      guard result == 0 else {
        continue
      }
      let address = String(cString: hostBuffer)
      if !address.isEmpty, !addresses.contains(address) {
        addresses.append(address)
      }
    }

    return addresses
  }
}
